File: implementation-nodejs/user-service/src/service.js
const userRepository = require('./repository');

class UserService
{
    async checkEligibility(userId, amountToSpend)
    {
        const user = await userRepository.findById(userId);
        if(!user)
            return { isValid: false }

        return {
            isValid: true,
            name: user.name,
            balanceSufficient: user.balance >= amountToSpend,
            currentBalance: user.balance
        }
    }

    async updateUserBalance(userId, amount, isCredit)
    {
        const user = await userRepository.findById(userId);
        if(!user)
            return { 
                status: false,
                message: 'User not found',
                balance: 0
            };

        if(!isCredit && user.balance < amount)
            return { 
                status: false,
                message: 'Insufficient balance for this transaction',
                balance: user.balance
            };

        const result = await userRepository.updateBalance(userId, amount, isCredit);
        if(!result.status)
            return { status: false, message: 'Failed to update balance', balance: user.balance };

        return { status: true, message: 'Balance updated successfully', balance: user.balance}
    }
}

module.exports = new UserService();

File: implementation-go/user-service/service.go
package user_service

import "log"

type UserService struct {
	repo *UserRepository
}

func (s *UserService) CheckUserEligibility(id int64, amount float64) (*User, bool, error) {
	user := s.repo.FindById(id)
	if user == nil {
		return nil, false, nil
	}

	return user, user.Balance > amount, nil
}

func (s *UserService) UpdateUserBalance(id int64, amount float64, isCredit bool) (status bool, message string, newBalance float64) {
	user := s.repo.FindById(id)
	if user == nil {
		return false, "User not found", 0
	}

	if !isCredit && user.Balance < amount {
		return false, "Insufficient balance for this transaction", user.Balance
	}

	status, newBalance = s.repo.UpdateBalance(id, amount, isCredit)
	if !status {
		return false, "Failed to update balance", user.Balance
	}

	return true, "Balance updated successfully", newBalance
}

func NewUserService() *UserService {
	return &UserService{
		repo: NewUserRepository(),
	}
}
